addons/pos_coupon/models/coupon_program.py

```diff
@@ -69,15 +69,25 @@ def action_view_pos_orders(self):
 
     @api.depends("rule_products_domain")
     def _compute_valid_product_ids(self):
+        domain_products = {}
         for program in self:
-            domain = ast.literal_eval(program.rule_products_domain) if program.rule_products_domain else []
-            program.valid_product_ids = self.env["product.product"].search(domain).ids
+            product_ids = domain_products.get(program.rule_products_domain)
+            if product_ids is None:
+                domain = ast.literal_eval(program.rule_products_domain) if program.rule_products_domain else []
```

Comment:
Maybe use `safe_eval` from `odoo.tools.safe_eval` instead of `ast.literal_safe`

Reply:
Changed! 

Thank you!

Reply:
The runbot is raising red status because of a security issue

Should we revert this security change?

Reply:
Ok my bad you can use literal_eval